execution/sym_region_frag_machine.ml

```diff
@@ -1253,22 +1272,44 @@ struct
 	else 
 	  match self#decide_wd "Load" off_exp cloc with
 	    | None -> None
-	    | Some wd -> self#table_load cloc off_exp wd ty
+	    | Some wd -> self#table_load cloc (Some 0) off_exp wd ty
 	    
     method private handle_load addr_e ty =
       if !opt_trace_offset_limit then
 	Printf.printf "Loading from... %s\n" (V.exp_to_string addr_e);
       match self#maybe_table_or_concrete_load addr_e ty with
-        | Some v -> (v, ty)
-        | None ->
-      let (r, addr) = self#eval_addr_exp_region addr_e 0x8000 in
-      let v =
-	(match ty with
-	   | V.REG_8  -> form_man#simplify8  (self#load_byte_region  r addr)
-	   | V.REG_16 -> form_man#simplify16 (self#load_short_region r addr)
-	   | V.REG_32 -> form_man#simplify32 (self#load_word_region  r addr)
-	   | V.REG_64 -> form_man#simplify64 (self#load_long_region  r addr)
-	   | _ -> failwith "Unsupported memory type") in
+      | Some v -> (v, ty)
+      | None ->
+	let (r, addr, is_table_load, off_expr, cloc ) = 
+	  self#eval_addr_exp_region addr_e 0x8000 in
+	let sym_region_table_v = 
+	  if is_table_load = true then (
+	    if addr <> 0L then 
+	      failwith "table load from symbolic region with non-zero offset, panic!";
+	    match self#decide_wd "Load" off_expr 0L with
+	    | None -> None
+	    | Some wd -> 
+	      Printf.printf 
+		"SRFM#handle_load table load for sym region with offset expr = %s\n"
+		(V.exp_to_string off_expr);
+	      self#table_load 0L r off_expr wd ty
+	  ) else (
+	    if cloc <> 0L then 
+	      failwith "table load from symbolic region with non-zero cloc, panic!";
```

Comment:
More generally, all the special cases about some of the data in the return tuple being present or not seems a bit error-prone. It occurs to me that a slightly more verbose but I think ML-idiomatic way to handle this kind of situation would be to use a datatype to represent the two different kinds of returns from sfrm#eval_addr_expr_region. It involves thinking up names but maybe it could be something like mem_location which is either a SingleLocation or a TableLocation. The SingleLocation just has an address, while the TableLocation has a cloc and an offset expression. The choice between the two types of location would play the role of the is_table_load boolean. (The way I just described it, the method would return a region and a mem_location; alternatively both mem_location cases could include the region.) The advantage of this approach is that the compiler can do more checking that the right data is available and both cases are always handled. You'd need to use a match every place we call SRFM#eval_addr_expr_region, but in most of those places you currently have to have an "if" to branch on is_table_load, so that would be more re-arranging complexity than adding it.